primitives/weights/src/weight_v2.rs

```diff
@@ -48,71 +51,118 @@ impl Weight {
 		self
 	}
 
+	/// Set the storage size part of the weight.
+	pub const fn set_proof_size(mut self, c: u64) -> Self {
+		self.proof_size = c;
+		self
+	}
+
 	/// Return the reference time part of the weight.
 	pub const fn ref_time(&self) -> u64 {
 		self.ref_time
 	}
 
-	/// Return a mutable reference time part of the weight.
+	/// Return the storage size part of the weight.
+	pub const fn proof_size(&self) -> u64 {
+		self.proof_size
+	}
+
+	/// Return a mutable reference to the reference time part of the weight.
 	pub fn ref_time_mut(&mut self) -> &mut u64 {
 		&mut self.ref_time
 	}
 
-	pub const MAX: Self = Self { ref_time: u64::MAX };
+	/// Return a mutable reference to the storage size part of the weight.
+	pub fn proof_size_mut(&mut self) -> &mut u64 {
+		&mut self.proof_size
+	}
+
+	pub const MAX: Self = Self { ref_time: u64::MAX, proof_size: u64::MAX };
 
 	/// Get the conservative min of `self` and `other` weight.
 	pub fn min(&self, other: Self) -> Self {
-		Self { ref_time: self.ref_time.min(other.ref_time) }
+		Self {
+			ref_time: self.ref_time.min(other.ref_time),
+			proof_size: self.proof_size.min(other.proof_size),
+		}
 	}
 
 	/// Get the aggressive max of `self` and `other` weight.
 	pub fn max(&self, other: Self) -> Self {
-		Self { ref_time: self.ref_time.max(other.ref_time) }
+		Self {
+			ref_time: self.ref_time.max(other.ref_time),
+			proof_size: self.proof_size.max(other.proof_size),
+		}
 	}
 
 	/// Try to add some `other` weight while upholding the `limit`.
 	pub fn try_add(&self, other: &Self, limit: &Self) -> Option<Self> {
 		let total = self.checked_add(other)?;
-		if total.ref_time > limit.ref_time {
+		if total.any_gt(limit) {
 			None
 		} else {
 			Some(total)
 		}
 	}
 
-	/// Construct [`Weight`] with reference time weight.
+	/// Construct [`Weight`] with reference time weight and 0 storage size weight.
 	pub const fn from_ref_time(ref_time: u64) -> Self {
-		Self { ref_time }
+		Self { ref_time, proof_size: 0 }
 	}
 
+	/// Construct [`Weight`] with storage size weight and 0 reference time weight.
+	pub const fn from_proof_size(proof_size: u64) -> Self {
+		Self { ref_time: 0, proof_size }
+	}
+
+	/// Construct [`Weight`] with weight components, namely reference time and storage size weights.
+	pub const fn from_components(ref_time: u64, proof_size: u64) -> Self {
+		Self { ref_time, proof_size }
+	} 
```

Comment:
this function will break if we add a third dimension in the future. does that imply we should not have it?

you could always do:

```
Weight::new().set_proof_size(a).set_ref_time(b)
```

Which would be equiv

Reply:
Well, it could be that it actually *should* break, to remind us to fix the code. Not sure…

Reply:
That's what I was thinking as well, the `from_components` method should break whenever there's a new field. I've wondered whether we should just exclusively use a builder pattern, but having the option to use just 1 single function to create a `Weight` type sounded appealing to me.

Reply:
i have no strong opinion in either direction. as long as we are being conscious of the choice i am happy :)